src/modules-by-example.adoc

```diff
@@ -0,0 +1,320 @@
+new shared symbol table.
+
+The latter imports-only document has this 1.1 equivalent:
+
+
+[{nrm}]
+----
+*$ion_1_1*
+**$ion_encoding**::{
+  modules: [
+    (*import* local "local" 1 2)
+  ],
+  install_symbols: [ local ]
+}
+----
+
+Here we see a new form inside the `modules` field that imports a module into the encoding
+environment and assigns it a name.
+The `*import*` keyword starts an S-expression that expects three or four arguments. The first is
+a symbolic name that we can use later to refer to the imported module.  The remaining arguments
+are effectively the `name`, `version` and `max_id` fields of the 1.0 `imports` struct.
+
+TIP: From the perspective of Ion 1.1, shared symbol tables _are_ encoding modules.
+
+
+=== Declaring Multiple Modules
+
+Within a directive's `modules` field, both `*module*` and `*import*` clauses have the same
+ultimate effect, declaring a name->module mapping for the encoding environment.
+Let’s look at a scenario with both imported and locally-defined symbols:
+
+[{nrm}]
+----
+*$ion_1_0*
+**$ion_symbol_table**::{
+  imports: [{ name: "shared1", version: 1, max_id: 10 },
+            { name: "shared2", version: 2, max_id: 20 }],
+  symbols: ["s1", "s2"]
+}
+----
+
+Here’s the Ion 1.1 equivalent in terms of symbol allocation order:
+
+[{nrm}]
+----
+*$ion_1_1*
+**$ion_encoding**::{
+  modules: [(*import* n1 "shared1" 1 10),
+            (*import* n2 "shared2" 2 20),
+            (*module* local (*symbols* ["s1", "s2"]))],
+  install_symbols: [n1, n2, local]
+}
+----
+
+Just as in the 1.0 version, this allocates symbol IDs starting with ten from `shared1`, twenty
+from `shared2`, then the two locally-defined symbols.
+
+By decoupling symbol-table importing from installation, Ion 1.1 allows some encoding techniques
+that are not possible in 1.0.  For example, we can give local symbols smaller IDs than imported
+symbols by installing `local` first:
+
+[{nrm}]
+----
+*$ion_1_1*
+**$ion_encoding**::{
+  modules: [(*import* n1 "shared1" 1 10),
+            (*import* n2 "shared2" 2 20),
+            (*module* local (*symbols* ["s1", "s2"]))]
+  install_symbols: [local, n1, n2]                       // 'local' is first
+}
+----
+
+
+=== Extending the Symbol Table
+
+The last 1.0 feature to examine is local symbol table extension:
+
+[{nrm}]
+----
+*$ion_1_0*
+**$ion_symbol_table**::{
+  symbols: ["s1", "s2"]
+}
+
+// ... application data ...
+
+**$ion_symbol_table**::{
+  imports: *$ion_symbol_table*,
+  symbols: ["s3", "s4"]
+}
+----
+
+To achieve this in Ion 1.1, we must copy the available modules from the current segment into
+the next, while also defining a new module for the additional symbols.
+
+[{nrm}]
+----
+*$ion_1_1*
+**$ion_encoding**::{
+  modules: [(*module* local (*symbols* ["s1", "s2"]))],
+  install_symbols: [local]
+}
+
+// ... application data ...
+
+**$ion_encoding**::{
+  modules: [
+    *{asterisk}*,                                      // retain all available modules
```

Comment:
This gets rendered as `*` right?

Reply:
Again, I was going to ask the same thing, but Almann is faster with the publish button. ;)

If it does render as __`*`__, is this special syntax or should it actually be single-quoted like a symbol? (If it's special syntax, we should call it out as such.)

Reply:
@popematt asks a really good question, `*` would not, unquoted, be valid _Ion_ syntax in a list here.  So is the syntax `'*'` or do we have a special form for this like `(*)` or `(inherit_context)` or whatever.

Reply:
Cut #206 to track that.

Reply:
What about `(module *)`? Or, I think that `(*)` could be a fine alternative, as Almann (sort of) suggested.

Reply:
Not discussed here is the ability to retain a selected set of available modules.  I'm inclined to go with clauses like `(retain *)` and `(retain mod1 mod3)` for this purpose.  I think a dedicated keyword is worthwhile since the behavior is so different from `import` and `module`.

Also not discussed here is the (tentative, TBD) ability to retain _installed_ modules in a directive, so they don't need to be explicitly named.  I've also used `*` there (in other docs) and it's similarly defective.